lib/streamlit/web/server/routes.py

```diff
@@ -59,28 +64,21 @@ def set_extra_headers(self, path: str) -> None:
         else:
             self.set_header("Cache-Control", "public")
 
-    def parse_url_path(self, url_path: str) -> str:
-        url_parts = url_path.split("/")
-
-        maybe_page_name = url_parts[0]
-        if maybe_page_name in self._pages:
-            # If we're trying to navigate to a page, we return "index.html"
-            # directly here instead of deferring to the superclass below after
-            # modifying the url_path. The reason why is that tornado handles
-            # requests to "directories" (which is what navigating to a page
-            # looks like) by appending a trailing '/' if there is none and
-            # redirecting.
-            #
-            # This would work, but it
-            #   * adds an unnecessary redirect+roundtrip
-            #   * adds a trailing '/' to the URL appearing in the browser, which
-            #     looks bad
-            if len(url_parts) == 1:
-                return "index.html"
-
-            url_path = "/".join(url_parts[1:])
-
-        return super().parse_url_path(url_path)
+    def validate_absolute_path(self, root: str, absolute_path: str) -> Optional[str]:
+        try:
+            return super().validate_absolute_path(root, absolute_path)
+        except tornado.web.HTTPError as e:
+            # If the file is not found, and there are no reserved paths,
+            # we try to serve the default file and allow the frontend to handle the issue.
+            if e.status_code == 404:
+                if any([self.path.endswith(x) for x in self._reserved_paths]):
+                    raise e
+
+                self.path = self.parse_url_path(self.default_filename or "index.html")
+                absolute_path = self.get_absolute_path(self.root, self.path)
+                return super().validate_absolute_path(root, absolute_path)
+
+            raise e
```

Comment:
@vdonato This might be the most controversial part. My assumption here is that we need these paths to actually 404 for the sake of identifying the correct path that provides the stream.

The rest can essentially be handled by the frontend, but let me know if my assumption is wrong.